### ISSUE TITLE
Redundant output projection

### ISSUE THREAD

**Issue by hans** [ytsvetko/TED#1]
@ytsvetko 
Still grokking this seq2seq code as I go.. it's pretty hairy!
As I was preparing code to sample from the decoder outputs, I think I discovered a duplicate output projection operation.

The first output projection happens in `attention_decoder`, where the output is a combined function of attention states and recurrent state: https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_functions.py#L286
This output value is the one that eventually gets passed around as the 'decoder output,' and it's the one that the label classifier uses to make its decisions. BUT it's not the one that the decoder uses to actually draw tokens!

Note that the `attention_decoder` calls the default `loop_function` at each timestep:
https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_functions.py#L269

And within the `loop_function` you are using, there is another output projection before the argmax is taken: https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_functions.py#L120
(In the current case, the `loop_function` is called with real output projection values, so this if-block applies.)

I think this could/would be a serious bug. Do you agree? I'll fix it on the `jon` branch for now..

**Comment by hans** [comment]
It might be the case that this was simply part of the TF seq2seq template.
Note how the code also projects the outputs using the output projection provided to the `loop_function` later in `Seq2SeqModel`: https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_model.py#L182

A proper label classifier would run *after* that projection rather than within the actual `model_with_buckets` function. I think I'll pull the label classifier out of the initial seq2seq call rather than removing the second output projection.

**Comment by hans** [comment]
This will be a nontrivial amount of refactoring, but it is necessary for the code to be correct, I think.

**Comment by hans** [comment]
Fixed on my branch, and I'll move forward with the RL now!

**Comment by ytsvetko** [comment]
I don't see any bugs there. This code is a fork of the TensorFlow seq2seq tutorial, which is quite hard to follow. There are at least 3 different places that perform output projection:

1. As part of the loss computation: https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_model.py#L113
2. In the loop_function, which only used when feed_previous is enabled (currently only in decoding, not in training): https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_functions.py#L120
3. To get the words used only in decoding, not in training: https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_model.py#L185

The label classification should be done in the embedding space and not on the one-hot values. Unfortunately it is not possible to do this such that we take a final decoder output (argmax) value, then perform an embedding lookup and then do label classifier. The embedding lookup is not differentiable, which will make it impossible to force the decoder to choose a different word.
As a workaround we perform the label classification on the decoder outputs prior to projection.

**Comment by hans** [comment]
Hi Yulia,

I think the problem is that the label classifier is one place you're not doing this output projection. It works directly off of `decoder_outputs`, which is actually not the final softmax value: https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_functions.py#L543

The `decoder_outputs` variable draws from this tensor, rather than the actual final output projection which is used to sample words: https://github.com/ytsvetko/TED/blob/c05b256d5d1ab5963fc38662ce8e6e61d9453599/src/seq2seq_functions.py#L286

So I agree with all of the points you made above, but this is still the spot where the projection is missing.
I've fixed this on my branch by simply moving the label classifier code into `Seq2SeqModel` and having it operate on `self.outputs` post-projection.
